### PR TITLE
Snorql Alerting

### PR DESCRIPTION
**Pull request by prajain12**
Signed-off-by: Prakul Jain <prakul.jain@udaan.com>

## Motivation and Context

Alerting feature is currently missing in Snorql. Alerting will enable users to get alerts when a certain threshold is crossed for a database.

## Description

1. Introduced alerting feature in Snorql
2. Added ActiveQueriesFilter Alert in snorql-extensions
3. Introduced ColumnName annotation in snorql-framework
4. Added a DataPurgingJob - Responsible to delete snorql persisted data at regular intervals
5. Restructured JobManager and its dependencies
6. Introduced abstract functions in Connection.kt and upstream classes

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Backward compatibility

- [ ] snorql-extensions
- [x] snorql-framework

## How Has This Been Tested?

Tested locally using SQL Dashboard.

## Migration and deployment details (if appropriate)

Changes in Framework. Will need documentation for Alerts and data purge jobs.

### CODE REVIEW

File: snorql-framework/src/main/kotlin/com/udaan/snorql/framework/alerts/AlertsJob.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import com.udaan.snorql.framework.models.*
+import org.quartz.Job
+import org.quartz.JobExecutionContext
+
+class AlertsJob<in T : AlertInput, O : IAlertResult, V : IAlertRecommendation> : Job {
+    private val logger = SqlMetricManager.logger
+
+    override fun execute(context: JobExecutionContext) {
+        try {
+//            logger.info("Inside AlertsJob Class")
```
Comment (lokeshdevnani) [inline snorql-framework/src/main/kotlin/com/udaan/snorql/framework/alerts/AlertsJob.kt]: delete

File: snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/alerts/alert/ActiveQueriesAlert.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                if (it.elapsedTime.toIntOrNull() != null) it.elapsedTime.toInt() >= alertInput.elapsedTimeThreshold
+                else false
+            }
+        }
+        if (alertInput.cpuTimeThreshold != null) {
+            filteredQueryList = filteredQueryList.filter {
+                if (it.cpuTime != null) it.cpuTime >= alertInput.cpuTimeThreshold
+                else false
+            }
+        }
```
Comment (lokeshdevnani) [inline snorql-extensions/src/main/kotlin/com/udaan/snorql/extensions/alerts/alert/ActiveQueriesAlert.kt:63]:
```suggestion
            filteredQueryList = filteredQueryList.filter {
               it.cpuTime != null && it.cpuTime >= alertInput.cpuTimeThreshold
            }
        }
```

**Review by lakshmi-udaan — COMMENTED** [review]
How are we deciding on the SNAPSHOT version?
Can we address the detekt comments & if not can we remove the unnecessary ones?

**Review by lokeshdevnani — CHANGES_REQUESTED** [review]

**Comment by prajain12** [comment]
> How are we deciding on the SNAPSHOT version? Can we address the detekt comments & if not can we remove the unnecessary ones?

SNAPSHOT version is a minor increase as these features will be rolled out in the next major release post incorporating all enhancements.
Working on detekt comments

**Review by lokeshdevnani — APPROVED** [review]
LGTM

**Pull request merged**